### ISSUE TITLE
Generate Space does not always work

### ISSUE THREAD

**Issue by SrNetoChan** [IfcOpenShell/IfcOpenShell#4162]
Creating spaces using the surrounding walls, using the generate space tool sometimes does not work as expected. By not working means it simply does not generate any ifcSpace but there's no error either. My first thinking was that somehow the wall were not closing, so I have make sure to extend all walls, by pressing shift+E selecing each pair of wall.

I have tested in blender 3 with the september version of blenderbim addon, but @CyrilWaechter tested with blender 4 using the lastest code.

**Comment by SrNetoChan** [comment]
Here's the project I was working on. Please forgive my inexperience both in blender and blenderbim. The project may not be very well organized.

I have created 4 spaces in a storie, but there are 3 other spaces where it's not working.

https://www.dropbox.com/scl/fi/5ofu1mmptas8doeuxea9h/modelo_casa_rua_sao_paulo.ifc?rlkey=npu8q10t04wl2fwo0zsf26v26&dl=0

**Comment by CyrilWaechter** [comment]
# Steps to reproduce
1. Open [modelo_casa_rua_sao_paulo_cleaned.zip](https://github.com/IfcOpenShell/IfcOpenShell/files/13813339/modelo_casa_rua_sao_paulo_cleaned.zip)
2. Select `Res do Chao` IfcBuildingStorey collection (probably not necessary)
3. Select 4 wall around a missing space like following
![Screenshot_20240102_203013](https://github.com/IfcOpenShell/IfcOpenShell/assets/9622862/ac632397-672c-4748-b0be-d134ab790038)
4. Press `Shift-A` shortcut to launch `Generate Spaces From Walls`
# Current behavior
No space created. No error raised.
# Expected behavior
Get a space created or get an explicit error message to explain why.
# System info
OS: Manjaro GNU/Linux
Softwares:
* Blender 4.0
* python 3.11.6
* blenderbim, fresh, on f478ec8d6b97b070923c188a003ad06b8e7bad9f

**Comment by maxfb87** [comment]
I'm trying to figure out the reason but without success...in the meantime you could try to use generate space from cursor position changing the RL parameter...with RL = 2 it works...

**Comment by maxfb87** [comment]
Ok so i figured out why... the reason is because there is a minimum of area in order to consider a space and now it is 3 square meters...
It was done in order to prevent the creation of impossible spaces but i think that 3 is too much conservative...maybe 1 can be enough, no?
Next days i'll do a PR that fixes it...

**Comment by SrNetoChan** [comment]
Or the house I am working on is too small :-P. But I guess that a service bathroom with 2 square meter is ok. I would leave it to the user to decide what can be considered a space or not. Some might want to consider a closet a space that they want to use for dynamic annotation.

**Comment by SrNetoChan** [comment]
Thank you for looking into it.

**Comment by theoryshaw** [comment]
> It was done in order to prevent the creation of impossible spaces but i think that 3 is too much conservative...maybe 1 can be enough, no?

I'd would vote for even smaller: 0.01 sm maybe.   For things like lining closets and shafts, etc.

**Comment by Moult** [comment]
Just a clarification here. There are two ways to generate spaces in the code right now:

 1. By positioning the 3D cursor and generating from that point (with an active storey).
 2. By selecting walls or other boundary elements and generating from those bounding walls.

Both methods fail in this scenario.

The first method takes everything "room bounding" in the storey, slices them at a sampling cutting plane, and takes that soup of lines and converts them into polygons. The polygon containing your cursor is extruded into your room. There is no "minimum area" in this situation.

It failed due to numerical tolerances. This fix is similar to how we've dealt with bugs related to  drawing fills. This has now been fixed.

The second method takes everything you've selected, gets all the possible polygons, then excludes any small polygons (I guess because they're less likely to be actual rooms? I haven't dug into the code too much).

I've reduced the min area threshold which now fixes it too.

It's kinda weird to have two methods to achieve the same thing. Over time I hope we can work out which approach is more robust (for 2.5D rooms at least) and consolidate.
